spring-ai-client-chat/src/main/java/org/springframework/ai/chat/client/advisor/AbstractConversationHistoryAdvisor.java

```diff
@@ -0,0 +1,54 @@
+package org.springframework.ai.chat.client.advisor;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.springframework.ai.chat.client.ChatClientRequest;
+import org.springframework.ai.chat.memory.ChatMemory;
+import org.springframework.ai.chat.messages.Message;
+
+/**
+ * Advisor for standard ChatMemory implementations
+ *
+ * @author Mark Pollack
+ * @since 1.0.0
+ */
+public abstract class AbstractConversationHistoryAdvisor extends AbstractChatMemoryAdvisor<ChatMemory> {
```

Comment:
I didn't fully understand the goal of this new class. In any case, I would use a different name because we're dealing with _chat memory_ and not _chat history_, which is something different (for convenience, I described the difference in the docs https://docs.spring.io/spring-ai/reference/api/chat-memory.html)

Reply:
I'm also wondering about whether we need all these abstract classes. We have a `BaseAdvisor` interface takes care of the blocking/non-blocking bit, so at least we can inherit from there (similar to what we did in `QuestionAnswerAdvisor`) and only have to implement `before()` and `after()`.

Having the same abstraction for vector store (dealing with text) and memory store (dealing with messages) might also be unnecessary. Since this is a breaking change, we might take a chance to review the design of this sets of advisors.

Reply:
Could we do the following to simplify them all?

- `MessageChatMemoryAdvisor implements BaseAdvisor`
- `PromptChatMemoryAdvisor implements BaseAdvisor`
- `VectorStoreChatMemoryAdvisor implements BaseAdvisor`

If needed, a `BaseChatMemoryAdvisor` interface (not an abstract class), implementing `BaseAdvisor`, could be introduced for sharing some logic between `MessageChatMemoryAdvisor` and `PromptChatMemoryAdvisor` since they both deal with messages. That would give the same flexibility for customisation, but it would make it easier for us to evolve those classes rather than having abstract/protected methods.

I would leave `VectorStoreChatMemoryAdvisor` on its own since it works very differently (embeddings instead of messages).

Reply:
Ok, thanks, let me look into that.  I was focused on removing duplication and having something more formal in the design to separate the message based retrieval MemoryAdvisors as compared to the VectorStore one that doesn't convert from Document to Messages.